### DISCUSSION TITLE
To import data from our MySQL DB into dynamics 365 using a PHP script

### DISCUSSION THREAD

**Discussion by sierratec** [AlexaCRM/dynamics-webapi-toolkit#62] [Q&A] [answered]
Hi

Our CRM version is 9.2.21052.00127

Our requirement is we need to import data from our MySQL DB into dynamics 365 leads using a PHP script.

We are able to connect  to 365 with post man

Please suggest do we have sample code

Thanks

**Comment by georged** [comment]
Hi @sierratec 

I suggest you look at the [tutorial](https://github.com/AlexaCRM/dynamics-webapi-toolkit/wiki/Tutorial) which does include some sample code. 

Thanks

**Reply by sierratec** [reply]
Hi,

We installed XAMPP 7.4.20 (PHP 7.4). We are using Windows 10.

We have provided here  the steps that we followed [https://www.sierratec.com/downloads/ref-16062021.docx](url). We got error as 'Class 'AlexaCRM\WebAPI\ClientFactory' not found'

Could you please suggest.

Thanks

**Reply by georged** [reply]
Sorry @sierratec  we're not going to download and look at a random Word document from the internets. Do you mind recording the steps here? You can include screenshots as well if you have to. When mentioning an error, please include the full stack.

Thanks
George

**Comment by sierratec** [comment]
Hi,
Please find the steps
1.	We installed ‘composer’
2.	Our ‘autoload.php’ is located in C:\xampp\phpMyAdmin\vendor
3.	We used below code. 
<?php
require_once '../../phpMyAdmin/vendor/autoload.php';

use \AlexaCRM\WebAPI;

$client = \AlexaCRM\WebAPI\ClientFactory::createOnlineClient(
    'https://xxxx.crm.dynamics.com',
    'xxxx',
    'xxxx'
);
?>
4.	But we are getting error as 
‘Fatal error: Uncaught Error: Class 'AlexaCRM\WebAPI\ClientFactory' not found in C:\xampp\htdocs\TestCRMImport\test.php:6 Stack trace: #0 {main} thrown in C:\xampp\htdocs\TestCRMImport\test.php on line 6’

Thanks

**Reply by georged** [reply]
Hi @sierratec 

1. There is no code in your post - can you please check?
2. C:\xampp\phpMyAdmin\vendor folder sounds like the wrong place - that's for phpMyAdmin.

What you need to do is to create a new folder, e.g. c:\xampp\sample, go to that folder and run 

`composer require alexacrm/dynamics-webapi-toolkit:dev-master`

in that folder. That will create a  subfolder `vendor` and `autoloader.php` in it. **That** is the autoloader you need to include using this code: 

```php
require_once 'vendor/autoload.php';
```

**Reply by sierratec** [reply]
Hi,
After executing this composer require alexacrm/dynamics-webapi-toolkit:dev-master inside my folder I am not facing issues. Will check further steps and update
Thanks

**Comment by georged** [comment, marked as answer]
This comment is the answer: https://github.com/AlexaCRM/dynamics-webapi-toolkit/discussions/62?_pjax=%23js-repo-pjax-container#discussioncomment-881985

**Reply by sierratec** [reply]
Hi,

**I am using the following code**

```
<?php
require_once 'vendor/autoload.php';

use \AlexaCRM\WebAPI;

$client = \AlexaCRM\WebAPI\ClientFactory::createOnlineClient(
    'https://domain.crm5.dynamics.com/',
    'xxxxxx',
    'yyyyyy'
);

$settings = new \AlexaCRM\WebAPI\OData\OnlineSettings();
$settings->instanceURI = 'https://domain.crm5.dynamics.com/';
$settings->applicationID = 'xxxxxx';
$settings->applicationSecret = 'yyyyyy';

$settings->tlsVerifyPeers = false;

$settings->apiVersion = '9.1';

$middleware = new \AlexaCRM\WebAPI\OData\OnlineAuthMiddleware( $settings );

$odataClient = new \AlexaCRM\WebAPI\OData\Client( $settings, $middleware );

$client = new \AlexaCRM\WebAPI\Client( $odataClient );

new \AlexaCRM\Xrm\ColumnSet( true );

$query = new \AlexaCRM\Xrm\Query\QueryByAttribute( 'contact' );
$query->AddAttributeValue( 'lastname', 'name' );
$query->AddOrder( 'firstname', \AlexaCRM\Xrm\Query\OrderType::Descending() );
$query->ColumnSet = new \AlexaCRM\Xrm\ColumnSet( [ 'fullname', 'emailaddress1' ] );
$collection = $client->RetrieveMultiple( $query );

var_dump($client);
?>
```

**I got the secret from here** 
[https://www.sierratec.com/downloads/ref-17062021-1.jpg](https://www.sierratec.com/downloads/ref-17062021-1.jpg)

**But I am getting below error**
Fatal error: Uncaught AlexaCRM\WebAPI\OData\AuthenticationException: Authentication at Azure AD failed. AADSTS7000215: Invalid client secret is provided. Trace ID: 6fbaa57e-6b58-43b1-b404-00dc75552c00 Correlation ID: fc43602e-ee71-47f2-8f28-77a12d639212 Timestamp: 2021-06-17 10:50:45Z in C:\xampp\htdocs\TestCRMImport\vendor\alexacrm\dynamics-webapi-toolkit\src\WebAPI\OData\OnlineAuthMiddleware.php:172 Stack trace: #0 C:\xampp\htdocs\TestCRMImport\vendor\alexacrm\dynamics-webapi-toolkit\src\WebAPI\OData\OnlineAuthMiddleware.php(207): AlexaCRM\WebAPI\OData\OnlineAuthMiddleware->acquireToken() #1 C:\xampp\htdocs\TestCRMImport\vendor\guzzlehttp\guzzle\src\PrepareBodyMiddleware.php(35): AlexaCRM\WebAPI\OData\OnlineAuthMiddleware::AlexaCRM\WebAPI\OData\{closure}(Object(GuzzleHttp\Psr7\Request), Array) #2 C:\xampp\htdocs\TestCRMImport\vendor\guzzlehttp\guzzle\src\Middleware.php(31): GuzzleHttp\PrepareBodyMiddleware->__invoke(Object(GuzzleHttp\Psr7\Request), Array) #3 C:\xampp\htdocs\TestCRMImport\vendor\guzzlehttp\guzzle\src\ in C:\xampp\htdocs\TestCRMImport\vendor\alexacrm\dynamics-webapi-toolkit\src\WebAPI\OData\OnlineAuthMiddleware.php on line 172

Could you please suggest.

Thanks

**Reply by georged** [reply]
@sierratec either your token is invalid or you did not create an app user in Dataverse mapped to the service principal
Most likely invalid token so regenerate it and copy *before* closing the window where it's displayed

**Reply by sierratec** [reply]
Hi,
Authentication is working now. I recreated the secret. But getting another error

Fatal error: Uncaught AlexaCRM\WebAPI\ToolkitException: Metadata request returned a 403 code in C:\xampp\htdocs\TestCRMImport\vendor\alexacrm\dynamics-webapi-toolkit\src\WebAPI\Client.php:488 Stack trace: #0 C:\xampp\htdocs\TestCRMImport\vendor\alexacrm\dynamics-webapi-toolkit\src\WebAPI\Client.php(275): AlexaCRM\WebAPI\Client->retrieveViaQueryByAttribute(Object(AlexaCRM\Xrm\Query\QueryByAttribute)) #1 C:\xampp\htdocs\TestCRMImport\test.php(33): AlexaCRM\WebAPI\Client->RetrieveMultiple(Object(AlexaCRM\Xrm\Query\QueryByAttribute)) #2 {main} thrown in C:\xampp\htdocs\TestCRMImport\vendor\alexacrm\dynamics-webapi-toolkit\src\WebAPI\Client.php on line 488

Could you please suggest.

Thanks

**Reply by georged** [reply]
@sierratec I suggest you try to follow steps again. It looks like you have not created a Dataverse user for the service principal you created, or you did not granted that user required roles. Hence the 403 error (access denied).

**Reply by sierratec** [reply]
Hi,
Are you referring this user [https://www.sierratec.com/downloads/ref-180621.jpg](https://www.sierratec.com/downloads/ref-180621.jpg). Could you please confirm
Thanks

**Reply by georged** [reply]
No. Please read [the instructions](https://docs.alexacrm.com/integration-cds/getting-started/#get-credentials).

**Reply by sierratec** [reply]
Hi,
403 error is fixed. I tried the following.
1.	Created a test ‘Contact’ via CRM interface and retrieved using PHP script  – This is working
2.	Created a test ‘Contact’ using PHP script – This is working
3.	Tried to view the created contact as in Sn.2  using CRM interface – Not able to view. 
4.	Tried to retrieve the created contact as in Sn.2 using PHP script – This is working

**Actually my account is system administrator account with Full Privilege. Still I am able to view only the created contacts and leads by me. Please suggest how to give permission for a user to view all posted contacts and leads**

Thank you

**Reply by georged** [reply]
@sierratec 

sys admin with the full access to data definitely has access to everything, guaranteed. You maybe looking at the wrong instance, or using the wrong view, or sys admin account is in administrative mode (no data access), etc. That sounds like pure UX issue, nothing to do with the toolkit that works correctly.

HTH
George

**Reply by sierratec** [reply]
Hi,
This is working now. Thanks for your time and support.
